### PR TITLE
Feature/ledger eth nonce signing

### PR DESCRIPTION
**Pull request by ccali11**
@shanejearley - Per your request, I combined `feature/ledger-eth` and `feature/nonce-signing` into one branch and am submitting this PR. Let me know if and when you want me to take a look at this with you or add any further clarification.

### CODE REVIEW

File: apps/web/src/composables/ledger.ts [hunk 1]
```diff
@@ -1,26 +1,72 @@
 import SpeculosHttpTransport from '@casimir/hw-transport-speculos'
-import TransportWebHID from '@ledgerhq/hw-transport-webhid'
+import TransportWebUSB from '@ledgerhq/hw-transport-webusb'
 import Eth from '@ledgerhq/hw-app-eth'
+import { ethers } from 'ethers'
 
 export default function useLedger() {
+  const bip32Path = "44'/60'/0'/0/0"
```
Comment (shanejearley) [inline apps/web/src/composables/ledger.ts:7]: Switch to `'44\'/60\'/0\'/0/0'`.

File: apps/web/src/composables/ledger.ts [hunk 2]
```diff
@@ -1,26 +1,72 @@
...
+  }
 
-    async function _getLedgerTransport() {
-        if (import.meta.env.PUBLIC_SPECULOS_PORT) {
-            return await SpeculosHttpTransport.open(`http://127.0.0.1:${import.meta.env.PUBLIC_SPECULOS_PORT}`)
-        } else {
-            return await TransportWebHID.create()
-        }
-    }
+  async function sendLedgerTransaction(transaction: any) {
```
Comment (shanejearley) [inline apps/web/src/composables/ledger.ts:24]: Add transaction type.

File: apps/web/src/composables/ledger.ts [hunk 3]
```diff
@@ -1,26 +1,72 @@
...
+    // TODO: Replace according to selected testnet
+    const infuraProvider = new ethers.providers.JsonRpcProvider(
```
Comment (shanejearley) [inline apps/web/src/composables/ledger.ts:26]: Refactor to use import.meta.env.ETHEREUM_RPC variable provided (will add to scripts in a commit here). Will be local 8545 for now, then `dev` env will use a testnet and `master` releases will use mainnet.

File: apps/web/src/composables/iopay.ts [hunk 4]
```diff
@@ -28,6 +28,10 @@ export default function useIoPay() {
     }
   }
 
+  const signIoTexMessage = async (message: string) => {
```
Comment (shanejearley) [inline apps/web/src/composables/iopay.ts:31]: Change to `signIoPayMessage` for consistency.

**Review by shanejearley — CHANGES_REQUESTED** [review]

File: apps/web/src/composables/ledger.ts (hunk 1 above)
Comment (ccali11) [reply apps/web/src/composables/ledger.ts:7]: Resolved!

File: apps/web/package.json [hunk 5]
```diff
@@ -10,11 +10,15 @@
   },
   "dependencies": {
     "@heroicons/vue": "^1.0.6",
+    "@ledgerhq/hw-app-btc": "^8.0.2",
     "@ledgerhq/hw-app-eth": "^6.29.4",
     "@ledgerhq/hw-transport-webhid": "^6.27.2",
+    "@ledgerhq/hw-transport-webusb": "^6.27.4",
+    "@ledgerhq/logs": "^6.10.0",
```
Comment (shanejearley) [inline apps/web/package.json:17]: Remove.

File: apps/web/package.json [hunk 6]
```diff
@@ -32,5 +36,8 @@
     "vite": "^2.9.9",
     "vite-plugin-pages": "^0.23.0",
     "vue-tsc": "^0.34.7"
+  },
```
Comment (shanejearley) [inline apps/web/package.json:39]: Remove.

File: apps/web/src/composables/ledger.ts (hunk 2 above)
Comment (shanejearley) [reply apps/web/src/composables/ledger.ts:24]: Use `ethers.utils.UnsignedTransaction`.

File: apps/web/src/composables/ledger.ts [hunk 7]
```diff
@@ -1,26 +1,72 @@
...
+      signature
+    )
+    const txHash = await infuraProvider.sendTransaction(signedTransaction)
+    console.log('txHash :>> ', txHash)
+  }
+
+  async function signMessageWithLedger(message: string) {
+    const _eth = await getLedgerEthSigner()
+    const signature = await _eth.signPersonalMessage(
+      "44'/60'/0'/0/0",
```
Comment (shanejearley) [inline apps/web/src/composables/ledger.ts:58]: Sign with the default constant with the same value.

File: apps/web/src/composables/ledger.ts (hunk 2 above)
Comment (ccali11) [reply apps/web/src/composables/ledger.ts:24]: Resolved.

File: apps/web/src/composables/iopay.ts (hunk 4 above)
Comment (ccali11) [reply apps/web/src/composables/iopay.ts:31]: Resolved.

File: apps/web/package.json [hunk 8]
```diff
@@ -10,11 +10,15 @@
   },
   "dependencies": {
     "@heroicons/vue": "^1.0.6",
+    "@ledgerhq/hw-app-btc": "^8.0.2",
```
[the remark this answers is not in the document]
Comment (ccali11) [reply apps/web/package.json:13]: Resolved.

File: apps/web/package.json [hunk 9]
```diff
@@ -10,11 +10,15 @@
...
     "@ledgerhq/hw-app-eth": "^6.29.4",
     "@ledgerhq/hw-transport-webhid": "^6.27.2",
+    "@ledgerhq/hw-transport-webusb": "^6.27.4",
+    "@ledgerhq/logs": "^6.10.0",
     "@walletconnect/client": "^1.8.0",
     "@walletconnect/qrcode-modal": "^1.8.0",
     "buffer": "^6.0.3",
+    "core-js": "^3.25.2",
```
[the remark this answers is not in the document]
Comment (ccali11) [reply apps/web/package.json:21]: Resolved.

File: apps/web/package.json (hunk 5 above)
Comment (ccali11) [reply apps/web/package.json:17]: Resolved.

File: apps/web/package.json (hunk 6 above)
Comment (ccali11) [reply apps/web/package.json:39]: Remove what?

Comment (shanejearley) [reply apps/web/package.json:39]: Sorry tagged the wrong line.

File: apps/web/package.json [hunk 10]
```diff
@@ -32,5 +36,8 @@
     "vite": "^2.9.9",
     "vite-plugin-pages": "^0.23.0",
     "vue-tsc": "^0.34.7"
+  },
+  "alias": {
```
Comment (shanejearley) [inline apps/web/package.json:40]: Remove alias.

**Review by shanejearley — CHANGES_REQUESTED** [review]

File: apps/web/package.json (hunk 10 above)
Comment (ccali11) [reply apps/web/package.json:40]: ah of course of course

File: apps/web/src/composables/ledger.ts (hunk 7 above)
Comment (ccali11) [reply apps/web/src/composables/ledger.ts:58]: I'm not sure what this means.

Comment (shanejearley) [reply apps/web/src/composables/ledger.ts:58]: Change line 58 to `bip32Path`.

File: apps/web/src/composables/ledger.ts (hunk 3 above)
Comment (shanejearley) [reply apps/web/src/composables/ledger.ts:26]: Going to pull down your recent changes after you push and finish this from the script side. After that, will also address the gas issue you ran into with sending the transaction.

Comment (ccali11) [reply apps/web/src/composables/ledger.ts]: @shanejearley - Sounds good. Won't resolve.

**Comment by ccali11** [comment]
@shanejearley - PR ready for merging once we address the conflicts, but given the files that have conflicts, I will wait for you to review and deconflict. I did the following since your last review:
    - [x] Review gas calculation (https://blockgeeks.com/guides/ethereum-gas/)
    - [ ] Consolidate all the math for calculating gas fees to a separate method
    - [x] Use the tx object as a consistent pattern for each wallet’s sendTransaction method
    - [x] Move disconnect WalletConnect method to wallet connect composable
    - [x] Move the ethers code to the ethers composable

After doing a bunch of background research on how gas works, I didn't end up making any changes to gas calculations in the ledger composable. We may want to address how we're doing that in a future PR (we can discuss the London upgrade and [EIP 1559](https://consensys.net/blog/quorum/what-is-eip-1559-how-will-it-change-ethereum/) at some point), but for now it seems to work fine.

I can't remember if you wanted to include the phantom improvements in this PR or not, but I'll just checkout a branch off of this one to add those, which should wrap up my "Close" tasks for the [week](https://github.com/consensusnetworks/casimir/discussions/123).

**Review by shanejearley — APPROVED** [review]

**Pull request merged**